lib/termtable.c

```diff
@@ -0,0 +1,487 @@
+/*
+ * ASCII table generator.
+ * Copyright (C) 2017  Cumulus Networks
+ * Quentin Young
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+#include <zebra.h>
+#include <stdio.h>
+
+#include "termtable.h"
+#include "memory.h"
+
+/* clang-format off */
+struct ttable_style ttable_styles[] = {
+	{	// default ascii
+		.corner = '+',
+		.rownums_on = false,
+		.indent = 1,
+		.border.top = '-',
+		.border.bottom = '-',
+		.border.left = '|',
+		.border.right = '|',
+		.border.top_on = true,
+		.border.bottom_on = true,
+		.border.left_on = true,
+		.border.right_on = true,
+		.cell.lpad = 1,
+		.cell.rpad = 1,
+		.cell.align = LEFT,
+		.cell.border.bottom = '-',
+		.cell.border.bottom_on = true,
+		.cell.border.top = '-',
+		.cell.border.top_on = false,
+		.cell.border.right = '|',
+		.cell.border.right_on = true,
+		.cell.border.left = '|',
+		.cell.border.left_on = false,
+	}, {	// blank, suitable for plaintext alignment
+		.corner = ' ',
+		.rownums_on = false,
+		.indent = 1,
+		.border.top = ' ',
+		.border.bottom = ' ',
+		.border.left = ' ',
+		.border.right = ' ',
+		.border.top_on = false,
+		.border.bottom_on = false,
+		.border.left_on = false,
+		.border.right_on = false,
+		.cell.lpad = 0,
+		.cell.rpad = 3,
+		.cell.align = LEFT,
+		.cell.border.bottom = ' ',
+		.cell.border.bottom_on = false,
+		.cell.border.top = ' ',
+		.cell.border.top_on = false,
+		.cell.border.right = ' ',
+		.cell.border.right_on = false,
+		.cell.border.left = ' ',
+		.cell.border.left_on = false,
+   }
+};
+/* clang-format on */
+
+void ttable_del(struct ttable *tt)
+{
+	for (int i = tt->nrows - 1; i >= 0; i--)
+		ttable_del_row(tt, i);
+
+	XFREE(MTYPE_TMP, tt->table);
+	XFREE(MTYPE_TMP, tt);
+}
+
+struct ttable *ttable_new(struct ttable_style *style)
+{
+	struct ttable *tt;
+
+	tt = XCALLOC(MTYPE_TMP, sizeof(struct ttable));
```

Comment:
Can you please add local `MTYPE`s for this with `DEFINE_MTYPE_STATIC`? MTYPE_TMP should really not see any new uses.